core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,303 @@
+package org.bykn.bosatsu
+
+import cats.data.{NonEmptyList, State}
+import cats.implicits._
+import rankn._
+
+import Identifier.Constructor
+
+sealed abstract class NormalExpression {
+  val maxLambdaVar: Option[Int]
+}
+
+object NormalExpression {
+  case class App(fn: NormalExpression, arg: NormalExpression)
+  extends NormalExpression {
+    val maxLambdaVar = (fn.maxLambdaVar.toList ++ arg.maxLambdaVar.toList)
+      .reduceLeftOption(Math.max)
+  }
+  case class ExternalVar(pack: PackageName, defName: Identifier)
+  extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Match(arg: NormalExpression,
+    branches: NonEmptyList[(Pattern[(PackageName, Constructor), Type], NormalExpression)])
+  extends NormalExpression {
+    val maxLambdaVar =
+      (arg.maxLambdaVar.toList ++ branches.toList.flatMap(_._2.maxLambdaVar))
+        .reduceLeftOption(Math.max)
+  }
+  case class LambdaVar(index: Int) extends NormalExpression {
+    val maxLambdaVar = Some(index)
+  }
+  case class Lambda(expr: NormalExpression) extends NormalExpression {
+    val maxLambdaVar = expr.maxLambdaVar.map(_ - 1).filter(_ >= 0)
+  }
+  case class Struct(enum: Int, args: List[NormalExpression])
+  extends NormalExpression {
+    val maxLambdaVar = args.flatMap(_.maxLambdaVar).reduceLeftOption(Math.max)
+  }
+  case class Literal(lit: Lit) extends NormalExpression {
+    val maxLambdaVar = None
+  }
+  case class Recursion(lambda: NormalExpression) {
+    val maxLambdaVar = None
+  }
+}
+
+object Normalization {
+  case class NormalExpressionTag(ne: NormalExpression, children: Set[NormalExpression])
```

Comment:
`children` is useful to know in caching for when an expression has the same normal expression as one of the expressions it links to. In this case we would only save the value from the linked to expression in the cache so that we don't end up with circular evaluations

Reply:
what is a child in this context?

Reply:
Is this basically a subgraph where `ne` points to all `children` forming an equivalence set?

If so, I wonder if another representation might be:
```
type NormTag = Long
type TagsToExpr = Map[NormTag, Set[NormalExpression]]
type ExprToTag = Map[NormalExpression, NormTag] // we may not need this since you are putting it in the tag on the TypedExpr?
```

Reply:
Thinking more, I wonder if we could get to a world where we make the `NormTag` a `Sha256HashValue` or something, where it is a hashed version of everything the expression depends on.

So, then, two expressions are the same (assuming no collisions in sha256) if they share the same value. That would be really cool since we could serialize these Normalization Tags and load them back up. You've probably thought about the pros and cons here. What do you think?

Reply:
Yeah... I think this is exactly right. For my purposes I was just using the case class instances as keys in a Map. The hashing would have to happen after the reductions of course, but that's an implementation detail.

That said, let me devil's advocate the other side.
a) (and I think this is a good argument for now) it's nice to have them unsha'd as we're trying to understand them and figure out how to use them. Eventually we can have tests where we take the sha of what were expecting. But while we're learning it's really nice for the failing test to give us what we expect
b) I think this is a little sillier of an argument, but in theory the normal expressions contain a bunch of static analysis that may be useful. For example, you could parse the normal expression at evaluation time and quickly see if a `Let` variable is used and if not just not evaluate. This is actually right in the `maxLambda` I think. The cool thing about this is it's a completely static decision so you don't need a `Promise` or a lazy `Eval`. You just skip it all together.

So my proposal would be not to sha these for now but set a calendar date to re-evaluate

Reply:
> what is a child in this context?

`children` is a flattened set of all the normalexpressions of all the child expressions (eg, any expressions linked to by this expression). It's useful for cases like

```
def foo(a):
  bar(a)
```
Since after reduction `foo` and `bar` have the same normal expression, if `foo` ends up getting stored in the cache then it is a function that passes `a` to a function that it will look up in the cache and get itself :scream:. This could really be replaced with a boolean.

Can you explain those `type`s more?

Reply:
I think we should schedule a video chat to talk about this. I'm still not sure I understand.

a type like `(A, Set[A])` looks like a node and a set of edges. So, a collection of those can be thought of as a graph. But is it a directed or undirected graph? The types don't say, but there might be the invariant that if edge `a -> b` exists so will `b -> a`.

Also, we can imagine another representation of a graph: `List[(A, A)]` which is the list of all edges. Lastly, I *think* the graphs we care about here are actually cliques: clusters of equivalent expressions, for those kinds of graphs, we can represent them as `(Set[A], Id)` some id type along with the set of nodes in the clique.

I still don't currently understand why we would choose one representation over the other. I think a comment would help clarify for me.